disabledPushdown -> pushdownLevel

cryogenian opened the pull request:
More pushdown control

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala:
```diff
@@ -20,14 +20,58 @@ import slamdata.Predef._
 import quasar.physical.mongo.MongoResource.{Database, Collection}
 
+trait PushdownLevel extends Product with Serializable
+final case object Disabled extends PushdownLevel
```
We've been defining case constructors in the parent type's companion object by convention.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoConfig.scala:
```diff
@@ -20,14 +20,58 @@ import slamdata.Predef._
+      case "disabled" => DecodeResult.ok(Disabled)
+      case "light" => DecodeResult.ok(Light)
+      case "full" => DecodeResult.ok(Full)
+      case _ => DecodeResult.fail("This is not PushdownLevel", CursorHistory(List()))
```
I think you can use the history from `j` here, instead of an empty one.

cryogenian commented:
Moved #104 into this pr to minimize merge problems.

+ Extracted `safeField` and `safeCartouches` 
+ Used history from outside json 
+ `PushdownLevel` extracted into separate file, moved case classes into `PusdownLevel` object 

[ch5592] [ch5588]

djspiewak commented:
@cryogenian Quick change to this menagerie of things. Can you default pushdown to **Off** for now and feature flag the drop-down selector in the frontend? We've had a request to make the rollout a little more controlled.

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/MongoExpression.scala:
```diff
@@ -140,8 +155,9 @@ object MongoExpression {
-      def eqv(a: Projection, b: Projection): Boolean = a.steps.toList === b.steps.toList
+    implicit val orderProjection: Order[Projection] = new Order[Projection] {
+      def compare(a: Projection, b: Projection): SInt =
+        Order[List[ProjectionStep]].compare(a.steps.toList, b.steps.toList)
```
:golf: `Order.orderBy(_.steps.toList)`

wemrysi commented on datasource/src/main/scala/quasar/physical/mongo/interpreter/Cartesian.scala:
```diff
@@ -22,18 +22,19 @@ import cats.syntax.traverse._
       uniqueKey: String,
       version: Version,
-      cartouches: Map[CPathField, (CPathField, List[ScalarStage.Focused])],
+      cartouches: Map[Field, (Field, List[ScalarStage.Focused])],
```
:+1:
